### PR TITLE
Retrieving the docs version number automatically

### LINKED ISSUE #334: Retrieving the docs version number automatically
Currently, the version number (`version = u'0.70'`) is hardcoded inside `doc/conf.py`. As we run `.pre_processing.py` prior to `sphinx-build` and therefore prior to any script relying on `doc/conf.py`, we could retrieve the version number from `include/mantella`, minimising the chance of errors and repetitions in our library.

The same goes for `README.md`.

### PR DESCRIPTION
**Pull request by dkiechle**
Fixes #334 

My contribution is licensed under the MIT license.

### CODE REVIEW

File: doc/.pre_processing.py [hunk 1]
```diff
@@ -288,8 +308,20 @@ class Colors:
...
   
   for change in changelog:
+    if change[1] == 3:
+      if not experimentalheader:
+        experimentalheader = True
+        changelogfile.write('\n.. list-table:: Experimental (functions are subject to change)\n')
+        changelogfile.write('  :widths: ' + str(first_column) + ' ' + str(100 - first_column) + '\n')
+        changelogfile.write('\n')
+        
+      changelogfile.write('  * - **Planned** till ' + str(change[0]) + '\n')
```
Comment (SebastianNiemann) [inline doc/.pre_processing.py]: For instead of till (we do not need to set us deadlines here ;))

File: doc/.pre_processing.py [hunk 2]
```diff
@@ -23,10 +23,16 @@ class Colors:
...
-if not os.path.isfile('./conf.py'):
+if not os.path.isfile('./.conf.py'):
   print("Could not find Sphinx's configuration file. Make sure to start this script within Mantella's documentation root path.")
   exit(1)
 
+# Setting the versions number for conf.py
+searchresult = re.search(r'MANTELLA_VERSION_MAJOR[ ]+(\d+)\n.*[ ](\d+)', open(''.join(glob.glob('../include/mantella[0-9]')), mode='r', encoding='utf-8').read())
+actualVersion = searchresult.group(1) + '.' + searchresult.group(2)
+conf_file = open('./.conf.py', mode='r', encoding='utf-8').read()
+open('./conf.py', mode='w', encoding='utf-8').write(re.sub(r'(project = .*)', '\\1\\nversion = u\'' + actualVersion + '\'' , conf_file, 0))
```
Comment (SebastianNiemann) [inline doc/.pre_processing.py]: Why do we need to do this? As `conf.py` is a python script in it self, couldn't `conf.py` also do it by itself?

File: doc/.pre_processing.py [hunk 3]
```diff
@@ -27,6 +27,12 @@ class Colors:
   print("Could not find Sphinx's configuration file. Make sure to start this script within Mantella's documentation root path.")
   exit(1)
 
+# Setting the versions number for conf.py
+searchresult = re.search(r'MANTELLA_VERSION_MAJOR[ ]+(\d+)\n.*[ ](\d+)', open(''.join(glob.glob('../include/mantella[0-9]')), mode='r', encoding='utf-8').read())
+actualVersion = searchresult.group(1) + '.' + searchresult.group(2)
```
Comment (SebastianNiemann) [inline doc/.pre_processing.py]: actual_version

File: doc/.pre_processing.py [hunk 4]
```diff
@@ -27,6 +27,12 @@ class Colors:
   print("Could not find Sphinx's configuration file. Make sure to start this script within Mantella's documentation root path.")
   exit(1)
 
+# Setting the versions number for conf.py
+searchresult = re.search(r'MANTELLA_VERSION_MAJOR[ ]+(\d+)\n.*[ ](\d+)', open(''.join(glob.glob('../include/mantella[0-9]')), mode='r', encoding='utf-8').read())
```
Comment (SebastianNiemann) [inline doc/.pre_processing.py]: search_result

File: doc/.pre_processing.py [hunk 5]
```diff
@@ -71,19 +79,31 @@ class Colors:
       print(' ' + Colors.ERROR + 'No versionadded tag found' + Colors.END)
       break
     else:
-      changelog = changelog + [(float(re.search(r'versionadded:: (\d.\d+).*', changes[1]).group(1)),2,changes[0])]
+      addedVersion = re.search(r'versionadded:: (\d.\d+).*', changes[1]).group(1)
```
Comment (SebastianNiemann) [inline doc/.pre_processing.py]: added_version

File: doc/.pre_processing.py [hunk 6]
```diff
@@ -71,19 +79,31 @@ class Colors:
...
+        changelog = changelog + [(float(addedVersion),3,changes[0])]
+        experimental = True
+      else:
+        changelog = changelog + [(float(addedVersion),2,changes[0])]
         
-    if 'versionchanged' in changes[1]:
-      for vchanges in re.findall(r'versionchanged:: (\d.\d+)\n[ ]+(.*)', comments):
-        changelog = changelog + [(float(vchanges[0]),1,changes[0],vchanges[1])]
+        if 'versionchanged' in changes[1]:
+          for vchanges in re.findall(r'versionchanged:: (\d.\d+)\n[ ]+(.*)', comments):
```
Comment (SebastianNiemann) [inline doc/.pre_processing.py]: version_changed

File: doc/.pre_processing.py [hunk 7]
```diff
@@ -288,8 +308,20 @@ class Colors:
   changelogfile.write('Changelog\n')
   changelogfile.write('=========\n')
   actualVersion = 0.0
+  experimentalheader = False
```
Comment (SebastianNiemann) [inline doc/.pre_processing.py]: add_experimental_header

File: doc/_themes/mantella/static/style.css [hunk 8]
```diff
@@ -89,6 +89,12 @@ nav > ul > li.current li a {
   color: #6b6b6b;
 }
 
+div.admonition.warning{
```
Comment (SebastianNiemann) [inline doc/_themes/mantella/static/style.css]: `main .warning {`

Comment (SebastianNiemann) [inline doc/_themes/mantella/static/style.css]: This should also be placed below 

```
main p {
  margin: 0 0 8px 0;
}
```

File: doc/conf.py [hunk 9]
```diff
@@ -5,7 +5,7 @@
 #
 
 project = u'Mantella'
-version = u'0.70'
+version = u'0.00'
```
Comment (SebastianNiemann) [inline doc/conf.py]: `0.0` no need to two zeros

File: doc/.pre_processing.py [hunk 10]
```diff
@@ -288,8 +308,20 @@ class Colors:
...
   
   for change in changelog:
+    if change[1] == 3:
+      if not experimentalheader:
+        experimentalheader = True
+        changelogfile.write('\n.. list-table:: Experimental (functions are subject to change)\n')
```
Comment (SebastianNiemann) [inline doc/.pre_processing.py]: changelogfile.write('\n.. list-table:: Experimental (functions are subject to change)')
changelogfile.write('\n  :widths: ' + str(first_column) + ' ' + str(100 - first_column))
changelogfile.write('\n')

File: doc/.pre_processing.py [hunk 11]
```diff
@@ -288,8 +308,20 @@ class Colors:
...
+        changelogfile.write('  :widths: ' + str(first_column) + ' ' + str(100 - first_column) + '\n')
+        changelogfile.write('\n')
+        
+      changelogfile.write('  * - **Planned** for ' + str(change[0]) + '\n')
```
Comment (SebastianNiemann) [inline doc/.pre_processing.py]: changelogfile.write('\n  * - **Planned** for ' + str(change[0]))
changelogfile.write('\n    - :cpp:any:`' + change[2] + '`)

File: doc/.pre_processing.py (hunk 6 above)
Comment (dkiechle) [inline doc/.pre_processing.py]: is for .. versionadded:: version

Comment (SebastianNiemann) [inline doc/.pre_processing.py]: Then `version_number` instead of `vchanges`

File: doc/.pre_processing.py [hunk 12]
```diff
@@ -287,11 +307,23 @@ class Colors:
...
   changelogfile.write('=========\n')
-  actualVersion = 0.0
+  actual_version = 0.0
+  add_experimental_header = False
   
   for change in changelog:
-    if actualVersion != change[0]:
-      actualVersion = change[0]
+    if change[1] == 3:
+      if not add_experimental_header:
```
Comment (SebastianNiemann) [inline doc/.pre_processing.py]: No the logic here becomes clear.

Please set `experimental` to True in the beginning, remove the `not` from the condition and set `experimental` to False afterwards.

**Pull request merged**